Implements a tooltip and dynamic selection checkbox

Louie808 opened the pull request:
# Ticket Link

[1463](https://uhawaii.atlassian.net/browse/GROUPINGS-1463)

# List of squashed commits

- Creates selectAll modal and functions
- Starts implementation of dynamic checkbox selection
- Finishes implementation of proper dynamic selection checkboxes
- Starts implementation of tooltip and checkbox into exclude tab
- Starts cleanup of code and removal of List selector
- Finishes cleanup of code for ticket-1463
- resolves merge conflicts
- Updates Jasmine tests
- Updates jasmine to use new variable names
- Minor code cleanup
- Fix Jasmine test and console errors

# Test Checklist

- [X] Exhibits Clear Code Structure:
- [X] Project Unit Tests Passed:
- [X] Project Jasmine Tests Passed:
- [X] Executes Expected Functionality:
- [X] Tested For Incorrect/Unexpected Inputs:

JorWo commented on src/main/java/edu/hawaii/its/groupings/controller/HomeController.java:
```diff
@@ -230,4 +231,9 @@ public String dynamicModal() {
         return "modal/groupingOwnersModal";
     }
+
+    @GetMapping(value = "/modal/{modalName}")
+    public String openModalPath(@PathVariable String modalName) {
+        return "modal/" + modalName;
+    }
```
Thanks for this addition. With this change I think we can remove all the modal functions above. Thanks.

JorWo replied:
Also make sure to update the tests for HomeController.

Louie808 replied:
Will this be another ticket? Removal of modal functions above this one that will replace, and update HomeController?

JorWo replied:
Yep, make another ticket for that.

JorWo commented on src/main/resources/templates/modal/selectAllModal.html:
Correct me if I am wrong, you mentioned on Slack that this modal is not used anymore, so can you remove this file? Thanks.

JorWo replied:
Correct me if I am wrong, it looks like selectAllModal was added in this ticket. If so, that means there is no where else where the selectAllModal exists.

Louie808 replied:
Yes it was added this ticket, and just planned to use the removal as easy ticket to navigate files.

JorWo replied:
In that case, sorry I changed my mind, let's just remove it here. I don't think there is a lot of value for a new member because they can simply use IntelliJ and search for the things to delete.

Louie808 replied:
I was going for creating a ticket for mentor to show their onboardee walk through to do git commands. But can remove it here.

JorWo replied:
Thanks.

mhodgesatuh commented:
Another ticket would be fine.



On Tue, Nov 14, 2023 at 12:30 AM Louie808 ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/main/java/edu/hawaii/its/groupings/controller/HomeController.java
> <https://urldefense.com/v3/__https://github.com/uhawaii-system-its-ti-iam/uh-groupings-ui/pull/872*discussion_r1392359309__;Iw!!PvDODwlR4mBZyAb0!TEHa3iAA26KqtTD97GQ88-idk14JNLGFADIdONcT1qZsZPwYIwuUeiTTxwz5VTgArQcfVr8p5i-S9nM8nzMLMgbGQA$>
> :
>
> > +    @GetMapping(value = "/modal/{modalName}")
> +    public String ***@***.*** String modalName) {
> +        return "modal/" + modalName;
> +    }
>
> Will this be another ticket? Removal of modal functions above this one
> that will replace, and update HomeController?
>
> —
> Reply to this email directly, view it on GitHub
> <https://urldefense.com/v3/__https://github.com/uhawaii-system-its-ti-iam/uh-groupings-ui/pull/872*discussion_r1392359309__;Iw!!PvDODwlR4mBZyAb0!TEHa3iAA26KqtTD97GQ88-idk14JNLGFADIdONcT1qZsZPwYIwuUeiTTxwz5VTgArQcfVr8p5i-S9nM8nzMLMgbGQA$>,
> or unsubscribe
> <https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/ACX2CMGDUZ222JVD4R7B4WTYENB4HAVCNFSM6AAAAAA7HVLZ22VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTOMRZGQYTCOBRGE__;!!PvDODwlR4mBZyAb0!TEHa3iAA26KqtTD97GQ88-idk14JNLGFADIdONcT1qZsZPwYIwuUeiTTxwz5VTgArQcfVr8p5i-S9nM8nzMQpg1cow$>
> .
> You are receiving this because your review was requested.Message ID:
> <uhawaii-system-its-ti-iam/uh-groupings-ui/pull/872/review/1729411811@
> github.com>
>

JorWo commented on src/main/resources/templates/fragments/include.html:
```diff
@@ -69,7 +72,8 @@ <h1 class="font-weight-bold text-dark pt-2 mb-0 d-inline-block">Include ({{group
                                ng-model="membersInCheckboxList[i.uhUuid]"
-                               ng-click="$event.stopPropagation()">
+                               id="{{'include-checkbox-' + i.uhUuid}}"
+                               ng-click="$event.stopPropagation();toggleCheckAllSelection('Include', 'single')">
```
Just a tiny change, could you add a space after the semi-colon? Thanks.

JorWo commented on src/main/resources/templates/fragments/include.html:
```diff
@@ -27,12 +24,18 @@ <h1 class="font-weight-bold text-dark pt-2 mb-0 d-inline-block">Include ({{group
                         <input class="position-relative ml-0" type="checkbox"
-                               aria-label="Check all checkboxes"
-                               ng-model="allSelected"
+                               aria-label="Check all in PAGE"
```
Let's change the aria-label to "Select all checkboxes in the current page" to make it more understandable for screen readers.

JorWo commented:
Could you replace all `function()` declarations with es6 style arrow functions? `() => {}`. Thanks!

JorWo commented on src/main/resources/templates/fragments/exclude.html:
```diff
@@ -62,10 +68,11 @@ <h1 class="font-weight-bold text-dark pt-2 mb-0 d-inline-block">Exclude ({{group
                                ng-model="membersInCheckboxList[e.uhUuid]"
-                               ng-click="$event.stopPropagation()">
+                               id="{{'exclude-checkbox-' + e.uhUuid}}"
+                               ng-click="$event.stopPropagation();toggleCheckAllSelection('Exclude', 'single')">
```
I find it a bit weird to see toggleCheckAllSelection being called for one of the singular checkboxes here. You might be able to simplify everything by keeping constant track of the number of checkboxes that are currently checked as a scope variable. 

In short, try to simplify the logic by using a scope variable to keep track of the total number of checkboxes currently selected. Next, use a second function maybe called updateTotalSelectedCheckboxes() that is appended to the ng-click so that every checkbox is able to change the total count. Overall, I want to be able to avoid using $scope.watch because it can get very laggy.

Louie808 replied:
I think I did attempt a implementation of keeping track of the total selected, but ran into a problem with total being triggered incorrectly. I can try again.

JorWo commented on src/main/resources/static/javascript/mainApp/grouping.controller.js:
```diff
@@ -70,7 +70,11 @@
+        // Checkbox list
         $scope.membersInCheckboxList = {};
+        $scope.paginatedPageChange = true;
+        $scope.showToolTip = false;
+        $scope.cbPageSelected = false;
```
Let's make some changes to the variable names
- I am pretty sure that is a variable already defined as $scope.showTooltip, I would prefer if we just put the conditional checked boxes > 20 right into the ng-show. 
- We can remove the cb prefix, I think it's clear enough to be just called $scope.pageSelected.

Louie808 commented:
> Could you replace all `function()` declarations with es6 style arrow functions? `() => {}`. Thanks!

Does this include all other functions implemented from before? Do I put it as const `function () = {}` or `$scope.function () = {}` for those other functions?

JorWo commented:
> > Could you replace all `function()` declarations with es6 style arrow functions? `() => {}`. Thanks!
> 
> Does this include all other functions implemented from before? Do I put it as `const function () = {}` or `$scope.function () = {}` for those other functions?

If you find any, yes. However, the only places that should stay as a function() are definitions of controllers because they require the use of `this`. Also `table.controller.js` has some `function()` declarations that cannot be converted to es6 arrow style.

Louie808 commented:
# Ticket Link

[1463](https://uhawaii.atlassian.net/browse/GROUPINGS-1463)

# List of squashed commits

- Addresses PR suggestions
- Refactors selectAll code
# Test Checklist

- [X] Exhibits Clear Code Structure:
- [X] Project Unit Tests Passed:
- [X] Project Jasmine Tests Passed:
- [X] Executes Expected Functionality:
- [X] Tested For Incorrect/Unexpected Inputs:
